File: Fundamentals_of_Software_Design/Python_language/Prefix_tree/prefix_tree.py
class TrieNode:
    __slots__ = ('children', 'data', 'has_data')
    def __init__(self):
        self.children = {}
        self.data = None
        self.has_data = False

class Trie:
    __slots__ = ('root', 'dtor')
    def __init__(self, dtor):
        self.root = TrieNode()
        self.dtor = dtor

def trie_create(dtor):
    return Trie(dtor)

def trie_free(tr):
    def _free_node(node):
        if node:
            for child in node.children.values():
                _free_node(child)
            if tr.dtor and node.has_data:
                tr.dtor(node.data)
    _free_node(tr.root)
    tr.root = None

def trie_has(tr, key):
    node = tr.root
    for char in key:
        if char not in node.children:
            return 0
        node = node.children[char]
    return 1 if node.has_data else 0

def trie_get(tr, key):
    node = tr.root
    for char in key:
        if char not in node.children:
            return None
        node = node.children[char]
    return node.data if node.has_data else None

def trie_set(tr, key, data):
    node = tr.root
    for char in key:
        if char not in node.children:
            node.children[char] = TrieNode()
        node = node.children[char]
    
    if tr.dtor and node.has_data:
        tr.dtor(node.data)
    
    node.data = data
    node.has_data = True
    return 1

def trie_update(tr, key, up, next):
    node = tr.root
    for char in key:
        if char not in node.children:
            node.children[char] = TrieNode()
        node = node.children[char]
    
    old_data = node.data if node.has_data else None
    node.data = up(old_data, next)
    node.has_data = True
    return 1

def _should_delete(node):
    return not node.has_data and not node.children

def trie_delete(tr, key):
    def _delete_recursive(node, key, index):
        if index == len(key):
            if not node.has_data:
                return False, False
            
            if tr.dtor:
                tr.dtor(node.data)
            node.data = None
            node.has_data = False
            return True, _should_delete(node)
        
        char = key[index]
        if char not in node.children:
            return False, False
        
        child = node.children[char]
        deleted, should_delete_child = _delete_recursive(child, key, index + 1)
        
        if should_delete_child:
            del node.children[char]
        
        return deleted, _should_delete(node)
    
    deleted, _ = _delete_recursive(tr.root, key, 0)
    return 1 if deleted else 0

def trie_traverse(tr, en, user):
    def _traverse_node(node, current_key):
        if node.has_data:
            en(current_key, node.data, user)
        
        for char, child in node.children.items():
            _traverse_node(child, current_key + char)
    
    _traverse_node(tr.root, "")

File: Fundamentals_of_Software_Design/C_language/Prefix_tree/prefix_tree.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include "prefix_tree.h"

#define ALPHABET_SIZE 256

typedef struct _TrieNode {
    struct _TrieNode* children[ALPHABET_SIZE];
    Pointer data;
    int is_end;
    char character;
} TrieNode;

static TrieNode* create_node(char ch) {
    TrieNode* node = (TrieNode*)calloc(1, sizeof(TrieNode));
    if (!node) return NULL;
    node->is_end = 0;
    node->data = NULL;
    node->character = ch;
    return node;
}

static void free_node(TrieNode* node, Destructor dtor) {
    if (!node) return;
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        if (node->children[i]) {
            free_node(node->children[i], dtor);
        }
    }
    if (node->is_end && dtor) {
        dtor(node->data);
    }
    free(node);
}

static TrieNode* find_node(const TrieNode* node, const char* key) {
    if (!node || !key) return NULL;
    TrieNode* current = (TrieNode*)node;
    for (size_t i = 0; key[i]; i++) {
        int ch = (unsigned char)key[i];
        if (!current->children[ch]) {
            return NULL;
        }
        current = current->children[ch];
    }
    return current->is_end ? current : NULL;
}

static void traverse_node(const TrieNode* node, char* buffer, size_t depth, Enumerator en, void* user) {
    if (!node) return;
    buffer[depth] = node->character;
    if (node->is_end) {
        buffer[depth + 1] = '\0';
        en(buffer, node->data, user);
    }
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        if (node->children[i]) {
            traverse_node(node->children[i], buffer, depth + 1, en, user);
        }
    }
}

Trie* trie_create(Destructor dtor) {
    Trie* trie = (Trie*)malloc(sizeof(Trie));
    if (!trie) return NULL;
    trie->root = create_node('\0');
    if (!trie->root) {
        free(trie);
        return NULL;
    }
    trie->dtor = dtor;
    trie->size = 0;
    return trie;
}

void trie_free(Trie* tr) {
    if (!tr) return;
    free_node(tr->root, tr->dtor);
    free(tr);
}

int trie_has(const Trie* tr, const char* key) {
    if (!tr || !key) return 0;
    return find_node(tr->root, key) != NULL;
}

Pointer trie_get(const Trie* tr, const char* key) {
    if (!tr || !key) return NULL;
    TrieNode* node = find_node(tr->root, key);
    return node ? node->data : NULL;
}

int trie_set(Trie* tr, const char* key, Pointer data) {
    if (!tr || !key) return 0;
    TrieNode* current = tr->root;
    for (size_t i = 0; key[i]; i++) {
        int ch = (unsigned char)key[i];
        if (!current->children[ch]) {
            current->children[ch] = create_node(key[i]);
            if (!current->children[ch]) return 0;
        }
        current = current->children[ch];
    }
    if (current->is_end && tr->dtor) {
        tr->dtor(current->data);
    }
    else if (!current->is_end) {
        tr->size++;
    }
    current->is_end = 1;
    current->data = data;
    return 1;
}

int trie_update(Trie* tr, const char* key, Updater up, Pointer next) {
    if (!tr || !key || !up) return 0;
    TrieNode* current = tr->root;
    for (size_t i = 0; key[i]; i++) {
        int ch = (unsigned char)key[i];
        if (!current->children[ch]) {
            current->children[ch] = create_node(key[i]);
            if (!current->children[ch]) return 0;
        }
        current = current->children[ch];
    }
    Pointer old_data = current->is_end ? current->data : NULL;
    Pointer new_data = up(old_data, next);
    if (!current->is_end) {
        tr->size++;
    }
    current->is_end = 1;
    current->data = new_data;
    return 1;
}

static int can_prune(TrieNode* node) {
    if (!node) return 1;
    if (node->is_end) return 0;
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        if (node->children[i]) {
            return 0;
        }
    }
    return 1;
}

static int delete_recursive(TrieNode* node, const char* key, size_t depth, Destructor dtor, int* deleted) {
    if (!node) return 0;
    if (key[depth] == '\0') {
        if (node->is_end) {
            if (dtor) {
                dtor(node->data);
            }
            node->is_end = 0;
            node->data = NULL;
            *deleted = 1;
            return can_prune(node);
        }
        return 0;
    }
    int ch = (unsigned char)key[depth];
    if (!node->children[ch]) {
        return 0;
    }
    int should_prune = delete_recursive(node->children[ch], key, depth + 1, dtor, deleted);
    if (should_prune) {
        free(node->children[ch]);
        node->children[ch] = NULL;
        return can_prune(node);
    }
    return 0;
}

int trie_delete(Trie* tr, const char* key) {
    if (!tr || !key) return 0;
    int deleted = 0;
    delete_recursive(tr->root, key, 0, tr->dtor, &deleted);
    if (deleted) {
        tr->size--;
    }
    return deleted;
}

void trie_traverse(const Trie* tr, Enumerator en, void* user) {
    if (!tr || !en) return;
    char buffer[1024];
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        if (tr->root->children[i]) {
            traverse_node(tr->root->children[i], buffer, 0, en, user);
        }
    }
}

size_t trie_size(const Trie* tr) {
    return tr ? tr->size : 0;
}
